include/mun/runtime.h

```diff
@@ -0,0 +1,275 @@
+            return std::equal(std::begin(arg_type.guid.b), std::end(arg_type.guid.b), std::begin(output_guid.b), std::end(output_guid.b));
+        }
+
+        template <typename Arg1, typename Arg2>
+        constexpr bool equal_args() {
+            auto arg1_guid = type_guid<Arg1>();
+            auto arg2_guid = type_guid<Arg2>();
+            return std::equal(std::begin(arg1_guid.b), std::end(arg1_guid.b), std::begin(arg2_guid.b), std::end(arg2_guid.b));
+        }
+
+        template <typename Output, typename... Args>
+        constexpr bool valid_fn(const MunFunctionSignature& signature) noexcept {
+            constexpr auto NUM_ARGS = sizeof...(Args);
+
+            if (signature.num_arg_types != NUM_ARGS) {
+                std::cerr << "Invalid number of arguments. Expected: " << std::to_string(NUM_ARGS) << ". Found: " << std::to_string(signature.num_arg_types) << "." << std::endl;
+                return false;
+            }
+
+            if constexpr (NUM_ARGS > 0) {
+                const MunTypeInfo* arg_ptr = signature.arg_types;
+                const bool valid_args[] = { valid_arg<Args>(*(arg_ptr++))... };
+                for (size_t idx = 0; idx < NUM_ARGS; ++idx) {
+                    if (!valid_args[idx]) {
+                        const char* arg_names[] = { type_name<Args>()... };
+                        std::cerr << "Invalid argument type at index " << idx << ". Expected: " << signature.arg_types[idx].name << ". Found: " << arg_names[idx] << "." << std::endl;
+                        return false;
+                    }
+                }
+            }
+
+            if (signature.return_type) {
+                if (!valid_arg<Output>(*signature.return_type)) {
+                    std::cerr << "Invalid return type. Expected: " << signature.return_type->name << ". Found: " << type_name<Output>() << "." << std::endl;
+                    return false;
+                }
+            } else if (!equal_args<void, Output>()) {
+                std::cerr << "Invalid return type. Expected: " << type_name<void>() << ". Found: " << type_name<Output>() << "." << std::endl;
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    // TODO: Automate the generation of GUIDs
+    template<typename T>
+    constexpr MunGuid type_guid() {
+        static_assert(false, "Unknown type")
+    }
+
+    template<typename T>
+    constexpr const char* type_name() {
+        static_assert(false, "Unknown type")
+    }
+
+    template<>
+    constexpr MunGuid type_guid<bool>() {
+        return MunGuid{
+            { 3, 250, 172, 250, 126, 222, 75, 18, 246, 18, 212, 156, 254, 63, 16, 230 }
+        };
+    }
+
+    template<>
+    constexpr const char* type_name<bool>() {
+        return "@core::bool";
+    }
+
+    template<>
+    constexpr MunGuid type_guid<double>() {
+        return MunGuid{
+            { 143, 142, 214, 135, 13, 131, 110, 33, 172, 233, 214, 19, 161, 21, 42, 130 }
```

Comment:
I prefer to write this as hex values to indicate that they are bytes.

Reply:
Changed.